faf-java-server-app/src/main/java/com/faforever/server/game/SpoofDetectorService.java

```diff
@@ -0,0 +1,130 @@
+
+    Optional<Player> reporteeOptional = playerService.getOnlinePlayer(reporteeId);
+    Requests.verify(reporteeOptional.isPresent(), ErrorCode.PLAYER_NOT_ONLINE, reporteeId);
+    Player reportee = reporteeOptional.get();
+
+    Game reporteeGame = reportee.getCurrentGame();
+    Requests.verify(reporteeGame != null, ErrorCode.OTHER_PLAYER_NOT_IN_GAME, reporteeId);
+
+    Requests.verify(reporterGame.equals(reporteeGame), ErrorCode.NOT_SAME_GAME, reportee);
+
+    boolean passesValidation = verifyName(reporter, name, reportee)
+      && verifyRating(reporter, mean, deviation, reportee, reporteeGame)
+      && verifyCountry(reporter, country, reportee)
+      && verifyAvatar(reporter, avatarUrl, avatarDescription, reportee);
+
+    if (!passesValidation) {
+      reportee.getFraudReportsByReporterId().put(
+        reporter.getId(),
+        new FraudReport(reporter.getId(), reporterGame.getId(), name, (int) mean, (int) deviation, country, avatarDescription, avatarUrl)
+      );
+
+      int reportedFrauds = reportee.getFraudReportsByReporterId().size();
+      if (reportedFrauds > 1 && reportedFrauds >= reporterGame.getConnectedPlayers().size() / 2) {
+        Collection<Player> peers = reporteeGame.getConnectedPlayers().values().stream()
+          .filter(player -> !Objects.equals(player.getId(), reporteeId))
+          .collect(Collectors.toList());
+
+        clientService.disconnectPlayerFromGame(reporteeId, peers);
+      }
+    }
+  }
+
+  private boolean verifyAvatar(Player reporter, String avatarUrl, String avatarDescription, Player reportee) {
+    Optional<Avatar> optionalAvatar = reportee.getAvailableAvatars().stream()
+      .filter(AvatarAssociation::isSelected)
+      .map(AvatarAssociation::getAvatar)
+      .findFirst();
+    if (optionalAvatar.isPresent()) {
+      Avatar avatar = optionalAvatar.get();
+      if (!Objects.equals(avatarUrl, avatar.getUrl())) {
+        log.debug("Avatar URL '{}' of player '{}' does not match in-game URL '{}' as reported by player '{}'",
+          avatarUrl, reportee, avatar.getUrl(), reporter);
+        return false;
+      }
+      if (!Objects.equals(avatarDescription, avatar.getDescription())) {
+        log.debug("Avatar description '{}' of player '{}' does not match in-game description '{}' as reported by player '{}'",
+          avatarDescription, reportee, avatar.getDescription(), reporter);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean verifyCountry(Player reporter, String country, Player reportee) {
+    if (!Objects.equals(country, reportee.getCountry())) {
+      log.debug("Country '{}' of player '{}' does not match in-game country '{}' as reported by player '{}'",
+        reportee.getCountry(), reportee, country, reporter);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean verifyRating(Player reporter, float mean, float deviation, Player reportee, Game reporteeGame) {
+    Rating reporteeRating;
```

Comment:
So the rating changes asynconiously, on the end of a game. What if the player says he has 1200 rating but at the second the check is done he has 1300 because previous game finished and he got rating????!!!!

Reply:
Actually there is little we can do here on Server side

Reply:
But it seems problematic to me

Reply:
That is acctaully not that unlikely it happened to me several times: I had a different rating then displayed i rejoin and it updates.